### PR TITLE
Fixed #24919 -- Allowed disabling of migrations on a per app basis

### PR DESCRIPTION
**Pull request by MarkusH**
https://code.djangoproject.com/ticket/24919

### CODE REVIEW

**Comment by andrewgodwin** [comment]
Looks good to me.

File: docs/ref/settings.txt [hunk 1]
```diff
@@ -1836,18 +1836,23 @@ Default::
...
     {'blog': 'blog.db_migrations'}
 
-In this case, migrations pertaining to the ``blog`` app will be contained in the ``blog.db_migrations`` package.
+In this case, migrations pertaining to the ``blog`` app will be contained in
+the ``blog.db_migrations`` package.
 
 If you provide the ``app_label`` argument, :djadmin:`makemigrations` will
 automatically create the package if it doesn't already exist.
 
+When you supply ``None`` as a value for an app Django will consider the app as
```
Comment (timgraham) [inline docs/ref/settings.txt]: app, Django
(also need a versionchanged annotation)
Could we include some motivation about the appropriate use cases?

**Comment by timgraham** [comment]
With `MIGRATION_MODULES = {'polls': None}`:
```
$ python manage.py makemigrations polls
Migrations for 'polls':
  0001_initial.py:
    - Create model Choice
    - Create model Poll
    - Add field poll to choice
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/home/tim/code/django/django/core/management/__init__.py", line 350, in execute_from_command_line
    utility.execute()
  File "/home/tim/code/django/django/core/management/__init__.py", line 342, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/tim/code/django/django/core/management/base.py", line 347, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/tim/code/django/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/home/tim/code/django/django/core/management/commands/makemigrations.py", line 150, in handle
    self.write_migration_files(changes)
  File "/home/tim/code/django/django/core/management/commands/makemigrations.py", line 169, in write_migration_files
    migrations_directory = os.path.dirname(writer.path)
  File "/home/tim/code/django/django/db/migrations/writer.py", line 300, in path
    return os.path.join(self.basedir, self.filename)
  File "/home/tim/code/django/django/db/migrations/writer.py", line 246, in basedir
    migrations_module = import_module(migrations_package_name)
  File "/home/tim/.virtualenvs/django34/lib/python3.4/importlib/__init__.py", line 100, in import_module
    if name.startswith('.'):
AttributeError: 'NoneType' object has no attribute 'startswith'
```
Should we have a friendlier message?

File: tests/migrations/test_loader.py [hunk 2]
```diff
@@ -179,6 +179,29 @@ def test_load_empty_dir(self):
...
+        self.assertEqual(migration_loader.migrated_apps, {'migrated_app'})
+        self.assertEqual(migration_loader.unmigrated_apps, set())
+
+    @override_settings(
+        INSTALLED_APPS=['migrations.migrations_test_apps.migrated_app'],
+        MIGRATION_MODULES={"migrated_app": None},
+    )
+    def test_marked_as_unmigrated(self):
+        """
+        Allows disabling of migrations for a particular app.
```
Comment (timgraham) [inline tests/migrations/test_loader.py]: MIGRATION_MODULES allows...

File: tests/migrations/test_loader.py [hunk 3]
```diff
@@ -179,6 +179,29 @@ def test_load_empty_dir(self):
                 "App missing __init__.py in migrations module not in unmigrated apps."
             )
 
+    @override_settings(
+        INSTALLED_APPS=['migrations.migrations_test_apps.migrated_app'],
+    )
+    def test_marked_as_migrated(self):
+        """
+        Allows disabling of migrations for a particular app.
```
Comment (timgraham) [inline tests/migrations/test_loader.py]: chop or change? (copied from other test)

File: docs/ref/settings.txt (hunk 1 above)
Comment (MarkusH) [inline docs/ref/settings.txt]: I'd rather not promote the feature too much. I wasn't even sure I wanted to document it at all.

Comment (timgraham) [inline docs/ref/settings.txt]: My opinion is that it's better to explain why it exists and discourage its use except for certain cases instead of leaving the reader wondering about it.

Comment (MarkusH) [inline docs/ref/settings.txt]: Alright, sounds reasonable. Here you go :smile:

**Comment by MarkusH** [comment]
Added a more friendly message for the `MigraitonWriter` failure.

File: django/db/migrations/writer.py [hunk 4]
```diff
@@ -241,6 +241,13 @@ def serialize_datetime(value):
     def basedir(self):
         migrations_package_name = MigrationLoader.migrations_module(self.migration.app_label)
 
+        if migrations_package_name is None:
+            raise ValueError(
+                "Django can't create migrations for app '%s' because "
+                "migrations have been disabled via the MIGRATION_MODULES "
+                "settings variable." % self.migration.app_label
```
Comment (timgraham) [inline django/db/migrations/writer.py]: "settings variable" -> "setting"

File: docs/ref/settings.txt [hunk 5]
```diff
@@ -1850,18 +1850,28 @@ Default::
...
+the ``blog.db_migrations`` package.
 
 If you provide the ``app_label`` argument, :djadmin:`makemigrations` will
 automatically create the package if it doesn't already exist.
 
+.. versionchanged:: 1.9
+
+When you supply ``None`` as a value for an app, Django will consider the app as
+an app without migrations regardless of an existing ``migrations`` submodule.
+This can for example be used to skip migrations for this app in tests. If this
```
Comment (timgraham) [inline docs/ref/settings.txt]: This can be used, for example, in a test settings file to skip migrations while testing (tables will still be created for the apps' models).

File: docs/ref/settings.txt [hunk 6]
```diff
@@ -1850,18 +1850,28 @@ Default::
...
     {'blog': 'blog.db_migrations'}
 
-In this case, migrations pertaining to the ``blog`` app will be contained in the ``blog.db_migrations`` package.
+In this case, migrations pertaining to the ``blog`` app will be contained in
+the ``blog.db_migrations`` package.
 
 If you provide the ``app_label`` argument, :djadmin:`makemigrations` will
 automatically create the package if it doesn't already exist.
 
+.. versionchanged:: 1.9
```
Comment (timgraham) [inline docs/ref/settings.txt]: versionadded might be more clear since `None` had no effect (as far as I know) in earlier versions

File: docs/ref/settings.txt [hunk 7]
```diff
@@ -1850,18 +1850,28 @@ Default::
...
+
+When you supply ``None`` as a value for an app, Django will consider the app as
+an app without migrations regardless of an existing ``migrations`` submodule.
+This can for example be used to skip migrations for this app in tests. If this
+is used in your general project settings, remember the
```
Comment (timgraham) [inline docs/ref/settings.txt]: remember to use the migrate --run-syncdb option if you want to create tables for the app.

File: tests/migrations/test_commands.py [hunk 8]
```diff
@@ -593,6 +593,21 @@ def test_makemigrations_empty_migration(self):
                 self.assertIn('dependencies=[\n]', content)
                 self.assertIn('operations=[\n]', content)
 
+    @override_settings(
```
Comment (timgraham) [inline tests/migrations/test_commands.py]: could be a single line

File: tests/migrations/test_commands.py [hunk 9]
```diff
@@ -593,6 +593,21 @@ def test_makemigrations_empty_migration(self):
...
+        MIGRATION_MODULES={"migrations": None},
+    )
+    def test_makemigrations_disabled_migrations_for_app(self):
+        """
+        Makes sure that makemigrations raises a nice error when migrations are
```
Comment (timgraham) [inline tests/migrations/test_commands.py]: I'd drop the "Makes sure that" preamble.

File: docs/ref/settings.txt [hunk 10]
```diff
@@ -1850,18 +1850,28 @@ Default::
 
     {}  # empty dictionary
 
-A dictionary specifying the package where migration modules can be found on a per-app basis. The default value
-of this setting is an empty dictionary, but the default package name for migration modules is ``migrations``.
+A dictionary specifying the package where migration modules can be found on a
```
Comment (timgraham) [inline docs/ref/settings.txt]: It's a bit nicer to do these cleanups in a separate commit if you wish.

File: docs/ref/settings.txt (hunk 6 above)
Comment (MarkusH) [inline docs/ref/settings.txt]: Yeah, it would just break.
